code-dot-org/code-dot-org
# allow lessons update server api to add/update/remove script levels from activity sections

FInishes [PLAT-280]. This PR is based off of https://github.com/code-dot-org/code-dot-org/pull/37146 which takes care of sending/displaying script level data on the lesson edit page.

What this PR does:
* supports add/update/delete of script levels on the server. This approach is mostly based off of https://github.com/code-dot-org/code-dot-org/pull/36935 which did the same thing for LessonActivity and ActivitySection, with ScriptLevels being nested one layer deeper, inside ActivitySection.
* fixes `position` and `chapter` on all script levels in the script via `fix_script_level_positions`. The strategy is to do this after all other updates are made to the lesson: https://github.com/code-dot-org/code-dot-org/blob/3da1240bb1881d07971421a6981232dec17d8e80/dashboard/app/models/lesson.rb#L421-L424
* best-effort attempt to translate server-side field names to client-side names and back

## Testing story

* new unit tests on fix_script_level_positions
* new controller and integration tests on the new add/update/delete operations on script levels
* manually verified that the lesson edit page can still save and load
* various protections in place to help avoid deleting script levels on existing scripts, though some of these are disabled in development

Here's what specifically is working (or not) in the lesson edit page now. I started with the valentine script, which looks like this on the script overview page:
![Screen Shot 2020-10-14 at 11 33 15 PM](https://user-images.githubusercontent.com/8001765/96085488-beccfe00-0e75-11eb-9625-ba60f68e8cd5.png)

To test out the lesson edit page, I used rails console to move all 4 levels on /s/valentine activity sections, which then looks like this:
![Screen Shot 2020-10-14 at 10 34 46 PM](https://user-images.githubusercontent.com/8001765/96084543-1c604b00-0e74-11eb-9f49-0f109a20013e.png)

Then I tried the following actions:
* add: not working, because the add level button doesn't add anything yet
* edit: works! I can mark a script level as an assessment on the lesson edit page, and the result shows up on the script overview page: 
![Screen Shot 2020-10-14 at 11 29 30 PM](https://user-images.githubusercontent.com/8001765/96085114-2b93c880-0e75-11eb-8452-cc0859ac07d8.png)
* reorder: not working, because the drag and drop isn't triggering the reorderLevel redux action. happy to chat about why @dmcavoy 
* delete: works! I deleted the first script level from the lesson edit page and you can see that they updated correctly on the script overview page:
![Screen Shot 2020-10-14 at 11 35 35 PM](https://user-images.githubusercontent.com/8001765/96085698-06538a00-0e76-11eb-8bfd-6178ca10ab17.png)

Caveat: the Lesson Overview page is still showing stub data, so you won't see any updates there.

# Reviewer Checklist:



[PLAT-280]: https://codedotorg.atlassian.net/browse/PLAT-280
